File: solutions/2026-06-11/java/2026-06-11-number-of-ways-to-assign-edge-weights-i.java
```java
// Problem: Number of Ways to Assign Edge Weights I
// Link: https://leetcode.com/problems/number-of-ways-to-assign-edge-weights-i/
// Approach:
// The problem asks for the number of ways to assign weights (1 or 2) to edges on the path from the root (node 1) to *any* node at the maximum depth, such that the total path cost is odd.
//
// First, we need to find the maximum depth of the tree and identify all nodes at that depth.
// For each node at the maximum depth, we need to consider the path from node 1 to it.
// The path from node 1 to any node 'x' in a tree is unique.
// Let the path from node 1 to node 'x' have 'k' edges. Each edge can be assigned a weight of 1 or 2.
// We need to find the number of ways to assign weights such that the sum of these 'k' weights is odd.
// The sum of weights is odd if and only if the number of edges assigned weight 1 is odd.
//
// Consider a path of length 'k' (k edges).
// For each edge, we have two choices: weight 1 or weight 2.
// Total number of ways to assign weights to 'k' edges is 2^k.
//
// Let 'o' be the number of edges assigned weight 1, and 't' be the number of edges assigned weight 2.
// Then o + t = k.
// The total cost is 1*o + 2*t.
// We want 1*o + 2*t to be odd.
// Since 2*t is always even, the parity of the sum depends solely on 'o'.
// So, we need 'o' (the number of edges with weight 1) to be odd.
//
// If k is the number of edges in the path:
// The number of ways to choose 'o' edges to have weight 1 is given by the binomial coefficient C(k, o).
// We need to sum C(k, o) for all odd values of 'o' from 1 to k.
// The sum of binomial coefficients for odd 'o' is C(k, 1) + C(k, 3) + C(k, 5) + ...
// It's a known mathematical identity that for any positive integer k,
// C(k, 0) + C(k, 1) + C(k, 2) + ... + C(k, k) = 2^k
// and
// C(k, 1) + C(k, 3) + C(k, 5) + ... = 2^(k-1)
// C(k, 0) + C(k, 2) + C(k, 4) + ... = 2^(k-1)
//
// So, for a path of length 'k' edges, there are exactly 2^(k-1) ways to assign weights such that the total cost is odd.
//
// The problem statement says: "Select *any one* node x at the maximum depth. Return the number of ways to assign edge weights in the path from node 1 to x such that its total cost is odd."
// This implies that we only need to consider *one* such node 'x' at the maximum depth.
// If there are multiple nodes at the maximum depth, they might have different path lengths from the root.
// However, the phrasing "Select any one node x at the maximum depth" suggests that we should pick *one* such node and calculate the ways for *that specific path*.
// If the problem intended the sum of ways for all such nodes, it would likely be phrased differently (e.g., "the sum of ways for all nodes at maximum depth").
// Let's re-read carefully: "Select any one node x at the maximum depth. Return the number of ways to assign edge weights in the path from node 1 to x such that its total cost is odd."
// This strongly suggests we need to find *a* node at max depth, and then calculate the ways for *its* path.
// The crucial part is that the tree is rooted at 1. The depth is implicitly defined from root 1.
//
// Algorithm:
// 1. Build an adjacency list representation of the tree.
// 2. Perform a Breadth-First Search (BFS) or Depth-First Search (DFS) starting from root node 1 to determine the depth of each node and the parent of each node. This will also help in identifying a node at maximum depth.
// 3. During the traversal (e.g., DFS), keep track of the current depth and the path length from the root.
// 4. Find the maximum depth reached.
// 5. Identify *one* node 'x' that is at this maximum depth. Since the tree is rooted at 1, any node at the maximum depth will have a unique path from the root. The length of this path (number of edges) is simply its depth minus 1.
// 6. Let 'max_depth' be the maximum depth found. A node 'x' at 'max_depth' will have a path of length 'max_depth - 1' edges from root 1.
// 7. The number of ways to assign weights to this path of length 'k = max_depth - 1' edges such that the total cost is odd is 2^(k-1).
// 8. The modulo is 10^9 + 7. We need to compute (2^(k-1)) % (10^9 + 7).
//
// Example 1: edges = [[1,2]]
// n = 2.
// Adjacency list: 1: [2], 2: [1].
// DFS from 1:
// - Visit 1, depth 0.
// - Visit 2 from 1, depth 1. Parent of 2 is 1.
// Max depth is 1. Node at max depth is 2.
// Path from 1 to 2 has 1 edge (1->2). k = 1.
// Number of ways = 2^(1-1) = 2^0 = 1.
//
// Example 2: edges = [[1,2],[1,3],[3,4],[3,5]]
// n = 5.
// Adjacency list: 1: [2,3], 2: [1], 3: [1,4,5], 4: [3], 5: [3].
// DFS from 1:
// - Visit 1, depth 0.
// - Visit 2 from 1, depth 1. Parent of 2 is 1.
// - Visit 3 from 1, depth 1. Parent of 3 is 1.
//   - Visit 4 from 3, depth 2. Parent of 4 is 3.
//   - Visit 5 from 3, depth 2. Parent of 5 is 3.
// Max depth is 2. Nodes at max depth are 4 and 5.
// Let's pick node 4.
// Path from 1 to 4: 1 -> 3 -> 4. Edges are (1,3) and (3,4). Length k = 2.
// Number of ways = 2^(2-1) = 2^1 = 2.
// If we pick node 5:
// Path from 1 to 5: 1 -> 3 -> 5. Edges are (1,3) and (3,5). Length k = 2.
// Number of ways = 2^(2-1) = 2^1 = 2.
// The problem states "Select *any one* node x at the maximum depth". This means we just need to calculate it for one such node. Since all nodes at the maximum depth will have paths of the same length (which is the maximum depth itself), the result will be the same. The length of the path from root 1 to a node at depth 'd' is 'd' edges if depth is 0-indexed, or 'd-1' edges if depth is 1-indexed. Given typical tree traversal, if root is depth 0, then a node at depth 'd' has 'd' edges from root. Let's use depth 0 for root. Then a node at max_depth 'D' has path length 'D'.
//
// Correction on path length: If root is depth 0, a node at depth `d` has `d` edges on the path from the root.
// Example 1: [[1,2]]. Node 2 is at depth 1. Path length = 1. k = 1. Ways = 2^(1-1) = 1.
// Example 2: [[1,2],[1,3],[3,4],[3,5]]. Max depth = 2. Nodes 4 and 5 are at depth 2. Path length = 2. k = 2. Ways = 2^(2-1) = 2.
//
// So, if max_depth is the maximum depth (root at depth 0), the path length from root to any node at max_depth is max_depth.
// We need to compute 2^(max_depth - 1) modulo 10^9 + 7.
// If max_depth is 0 (only root exists, n=1 which is not possible by constraints 2<=n), it's an edge case. But n>=2, so max_depth will be at least 1.
// If max_depth = 1 (e.g., [[1,2]]), path length = 1. k=1. Ways = 2^(1-1) = 2^0 = 1.
//
// Implementation details:
// - Adjacency list: `List<List<Integer>> adj`
// - Keep track of visited nodes in DFS/BFS.
// - Keep track of depth and parent during DFS.
// - `maxDepth` variable to store the maximum depth found.
// - `nodeAtMaxDepth` to store one such node (not strictly needed if we only care about maxDepth value).
// - A power function to compute (base^exp) % mod.
//
// Let's refine the depth and path length calculation.
// If root (node 1) is at depth 0.
// Its children are at depth 1.
// A node at depth `d` has `d` edges from the root.
// So, if `max_depth` is the maximum depth found by DFS/BFS, then the path length to any node at `max_depth` is `max_depth` edges.
// Let path length `k = max_depth`.
// Number of ways = 2^(k-1).
//
// Example 1: [[1,2]]
// Root 1 at depth 0.
// Node 2 at depth 1.
// max_depth = 1.
// Path length k = max_depth = 1.
// Ways = 2^(1-1) = 2^0 = 1.
//
// Example 2: [[1,2],[1,3],[3,4],[3,5]]
// Root 1 at depth 0.
// Nodes 2, 3 at depth 1.
// Nodes 4, 5 at depth 2.
// max_depth = 2.
// Path length k = max_depth = 2.
// Ways = 2^(2-1) = 2^1 = 2.
//
// This seems consistent.
// The constraints mention node labels from 1 to n. Our adjacency list and visited arrays should handle this.
// We can use 1-based indexing for adjacency list and visited arrays, or convert to 0-based internally. Using 1-based is often simpler if the problem uses it.
//
// `MOD = 10^9 + 7`
//
// DFS approach to find max depth:
// `dfs(currentNode, currentDepth, parentNode, adj, maxDepthTracker)`
// `maxDepthTracker` can be an array of size 1 or a custom object to pass by reference.
//
// `visited` array to prevent cycles (though it's a tree, it's good practice for graph traversals).
//
// Let's define depth 0 for the root.
// When we traverse from node `u` to `v` at `currentDepth + 1`, the path length from root to `v` is `currentDepth + 1`.
//
// Function `power(base, exp)`: computes `(base^exp) % MOD`.
// Standard modular exponentiation.
//
// Time Complexity:
// - Building adjacency list: O(N) where N is the number of nodes.
// - DFS to find maximum depth: O(N + E), where E is the number of edges. Since it's a tree, E = N-1. So, O(N).
// - Modular exponentiation: O(log K) where K is the exponent (max_depth). Since max_depth can be at most N-1, this is O(log N).
// Overall Time Complexity: O(N).
//
// Space Complexity:
// - Adjacency list: O(N + E) = O(N).
// - Visited array, depth array, recursion stack for DFS: O(N).
// Overall Space Complexity: O(N).

class Solution {
    long MOD = 1_000_000_007;
    // Adjacency list to store the tree
    List<List<Integer>> adj;
    // To store the maximum depth found during DFS
    int maxDepth;

    public int numberOfWaysToAssignEdgeWeights(int[][] edges) {
        int n = edges.length + 1; // Number of nodes is number of edges + 1

        // Initialize adjacency list
        adj = new ArrayList<>(n + 1); // Using 1-based indexing
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        // Build the adjacency list from the given edges
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        // Initialize maxDepth to -1 (or 0, if root is depth 0)
        // Since n >= 2, there will be at least one edge and depth >= 1 for children.
        maxDepth = 0;

        // Perform DFS starting from root node 1 to find the maximum depth.
        // We pass 0 for initial depth and -1 for parent to indicate root.
        dfs(1, 0, -1);

        // The problem asks for the number of ways to assign weights to edges
        // on the path from node 1 to *any* node 'x' at the maximum depth,
        // such that the total cost of the path is odd.
        //
        // If the root is at depth 0, a node at depth 'd' has 'd' edges on its path from the root.
        // So, if the maximum depth found is 'maxDepth', the path length to any node at maxDepth is 'maxDepth' edges.
        // Let 'k' be the number of edges in the path from root to a node at maximum depth. So, k = maxDepth.
        //
        // Each edge can be assigned a weight of 1 or 2.
        // The total cost is the sum of these weights.
        // The sum is odd if and only if the number of edges with weight 1 is odd.
        //
        // For a path with 'k' edges, the number of ways to assign weights such that the path cost is odd is 2^(k-1).
        // This is a known combinatorial result: the sum of C(k, i) for all odd 'i' is 2^(k-1).
        //
        // The constraint n >= 2 guarantees that there's at least one edge, so maxDepth will be at least 1.
        // Thus, k = maxDepth >= 1. The exponent k-1 will be >= 0.
        int pathLength = maxDepth; // Number of edges in path to a node at maxDepth

        // If pathLength is 0 (which means only root exists, but n>=2, so this won't happen),
        // the problem constraints prevent this. If pathLength is 1 (e.g., [[1,2]]), exponent is 0.
        // If pathLength is 0, 2^(-1) is undefined, but this scenario is ruled out.
        if (pathLength == 0) {
            // This case should not happen given n >= 2. If it did, it might imply 0 ways
            // or 1 way (if root itself could be considered at max depth, path length 0, cost 0, even)
            // But the problem states n >= 2, so maxDepth is at least 1.
            // For n=2, edges=[[1,2]], maxDepth=1, pathLength=1.
            return 0; // Or handle as per problem spec if this case was possible
        }

        // The number of ways is 2^(pathLength - 1) modulo MOD.
        // We need to compute (2 raised to the power of (pathLength - 1)) % MOD.
        // We use modular exponentiation.
        return (int) power(2, pathLength - 1);
    }

    /**
     * Depth-First Search to traverse the tree and find the maximum depth.
     * @param u The current node.
     * @param depth The current depth of node 'u' from the root (node 1). Root is at depth 0.
     * @param parent The parent of the current node 'u'. Used to avoid going back up the tree.
     */
    private void dfs(int u, int depth, int parent) {
        // Update the maximum depth found so far.
        // The depth of the current node 'u' is 'depth'.
        maxDepth = Math.max(maxDepth, depth);

        // Iterate through all neighbors of the current node 'u'.
        for (int v : adj.get(u)) {
            // If the neighbor 'v' is not the parent node (to avoid traversing back up),
            // then it's a child node.
            if (v != parent) {
                // Recursively call DFS for the child node 'v'.
                // The depth of 'v' is one greater than the depth of 'u'.
                dfs(v, depth + 1, u);
            }
        }
    }

    /**
     * Computes (base^exp) % mod using modular exponentiation (binary exponentiation).
     * This is efficient for large exponents.
     * @param base The base of the exponentiation.
     * @param exp The exponent.
     * @return (base^exp) % MOD.
     */
    private long power(long base, int exp) {
        long res = 1;
        base %= MOD; // Ensure base is within modulo range

        while (exp > 0) {
            // If exp is odd, multiply the result by the current base.
            if (exp % 2 == 1) {
                res = (res * base) % MOD;
            }
            // Square the base and halve the exponent for the next iteration.
            base = (base * base) % MOD;
            exp /= 2;
        }
        return res;
    }
}
```

File: solutions/2026-06-11/cpp/2026-06-11-number-of-ways-to-assign-edge-weights-i.cpp
```cpp
// Problem Summary: Assign weights (1 or 2) to edges in a tree such that the path cost from root (node 1) to a max-depth node is odd. Return the count modulo 10^9 + 7.
// Link: https://leetcode.com/problems/number-of-ways-to-assign-edge-weights-i/
// Approach:
// 1. The problem asks for the number of ways to assign edge weights (1 or 2) to edges along a path from the root (node 1) to *any* node at the maximum depth, such that the total path cost is odd.
// 2. A path cost is odd if the sum of edge weights is odd. This happens if there's an odd number of edges with weight 1.
// 3. The crucial observation is that for any node `x` at maximum depth, the path from root `1` to `x` has a fixed number of edges, say `k`.
// 4. For each edge in this path, we can assign weight 1 or 2. The total cost will be `w_1 + w_2 + ... + w_k`, where `w_i` is either 1 or 2.
// 5. The parity of the sum depends on the parity of the weights. Specifically, the sum is odd if and only if the number of edges with weight 1 is odd.
// 6. If an edge has weight 1, its contribution to the sum's parity is 1. If it has weight 2, its contribution is 0 (even).
// 7. So, for a path of length `k`, we need to choose a subset of edges to have weight 1, such that the size of this subset is odd.
// 8. The number of ways to choose an odd number of items from `k` items is `C(k, 1) + C(k, 3) + C(k, 5) + ...`. This sum is equal to `2^(k-1)` for `k >= 1`.
// 9. If `k=0` (i.e., the root itself is the max depth node), the path has 0 edges, cost is 0 (even). No ways to make it odd. But `n>=2` so this case won't happen for max depth unless n=1, which is ruled out.
// 10. We need to find a node `x` at the maximum depth. Since the problem states "Select any one node x at the maximum depth", and the result should be the same for any such node, we can find one such node and calculate the path length to it.
// 11. A simple DFS or BFS can be used to find the maximum depth and the path length to a node at that depth. We can perform a DFS from the root (node 1) to compute the depth of each node and keep track of the maximum depth encountered and the path length to any node at that maximum depth.
// 12. During DFS, we can pass the current depth and the current path length (number of edges from root) to the recursive calls.
// 13. When we reach a leaf node, we compare its depth with the current maximum depth.
//     - If it's greater than the current maximum depth, we update the maximum depth and the path length to this node.
//     - If it's equal to the current maximum depth, it means we found another node at the same maximum depth. Since the problem implies the answer is unique for any such node, we don't need to do anything specific here (or we could update if needed, but it's stated "any one node x"). The number of ways for any max depth node is the same.
// 14. Once the DFS is complete, we will have the path length `k` to *a* node at the maximum depth.
// 15. The answer is `2^(k-1)` modulo 10^9 + 7. If `k=0` (which won't happen for `n>=2` and max depth node not root), the answer would be 0. For `k>=1`, it's `2^(k-1)`.

#include <vector>
#include <queue>
#include <cmath>
#include <algorithm>

using namespace std;

class Solution {
    // Adjacency list to represent the tree
    vector<vector<int>> adj;
    // To store the maximum depth found
    int maxDepth;
    // To store the path length (number of edges) to a node at max depth
    int pathToMaxDepthNode;
    // Modulo constant
    long long MOD = 1e9 + 7;

    // Helper function for DFS to find max depth and path length
    void dfs(int u, int parent, int currentDepth, int currentPathLength) {
        // Update maxDepth and pathToMaxDepthNode if we find a deeper node
        if (currentDepth > maxDepth) {
            maxDepth = currentDepth;
            pathToMaxDepthNode = currentPathLength;
        }

        // If current node is a leaf (and not the root itself, which is handled implicitly by not calling DFS on parent)
        // Note: In a tree, a leaf is a node with only one neighbor, unless it's the root and n=1.
        // For n>=2, the root will have at least one child.
        // A simpler check for leaf in DFS is if it has no unvisited children other than its parent.
        // Here, we check if we've reached a potentially new max depth or an existing max depth.
        // The key logic is that any node at maxDepth will be a "terminal" node for the path calculation.

        // Explore neighbors
        for (int v : adj[u]) {
            if (v != parent) {
                // Recurse with incremented depth and path length
                dfs(v, u, currentDepth + 1, currentPathLength + 1);
            }
        }
    }

    // Function to calculate (base^exp) % mod
    long long power(long long base, long long exp) {
        long long res = 1;
        base %= MOD;
        while (exp > 0) {
            if (exp % 2 == 1) res = (res * base) % MOD;
            base = (base * base) % MOD;
            exp /= 2;
        }
        return res;
    }

public:
    int numberOfWaysToAssignEdgeWeights(vector<vector<int>>& edges) {
        // Number of nodes
        int n = edges.size() + 1;

        // Initialize adjacency list
        adj.resize(n + 1);
        for (const auto& edge : edges) {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        // Initialize variables for DFS
        maxDepth = 0;
        pathToMaxDepthNode = 0;

        // Start DFS from root (node 1)
        // Root is at depth 0, path length 0.
        // We are looking for the path length to a node at max depth.
        // The problem statement implies node 1 is the root, and paths are from node 1.
        dfs(1, 0, 0, 0);

        // The path length to the maximum depth node is pathToMaxDepthNode.
        // Let this path length be k.
        // We need to assign weights (1 or 2) to these k edges.
        // The total cost is odd if the number of edges assigned weight 1 is odd.
        // For k edges, the number of ways to choose an odd number of edges to have weight 1 is:
        // C(k, 1) + C(k, 3) + C(k, 5) + ... = 2^(k-1) for k >= 1.
        // If k = 0 (only possible if n=1 and root is max depth, which is not the case here since n>=2), the cost is 0 (even), so 0 ways.
        // Since n >= 2, the root is not the maximum depth node if it's the only node.
        // If the root is the only node at maximum depth (e.g., a star graph where n=2, edges=[[1,2]], root 1 is depth 0, node 2 is depth 1. Max depth is 1), pathToMaxDepthNode will be 1.
        // If n=2, edges=[[1,2]]. dfs(1,0,0,0) -> dfs(2,1,1,1). maxDepth=1, pathToMaxDepthNode=1. Result 2^(1-1) = 2^0 = 1. Correct.
        // If n=3, edges=[[1,2],[1,3]]. dfs(1,0,0,0) -> dfs(2,1,1,1), dfs(3,1,1,1). maxDepth=1, pathToMaxDepthNode=1 (either path length to 2 or 3). Result 2^(1-1) = 1. Correct.
        // If n=4, edges=[[1,2],[1,3],[3,4]]. dfs(1,0,0,0) -> dfs(2,1,1,1), dfs(3,1,1,1) -> dfs(4,3,2,2). maxDepth=2, pathToMaxDepthNode=2 (path 1->3->4). Result 2^(2-1) = 2^1 = 2. Correct.

        // If pathToMaxDepthNode is 0 (only if n=1, but n>=2), the answer is 0.
        // Otherwise, it's 2^(pathToMaxDepthNode - 1) mod MOD.
        if (pathToMaxDepthNode == 0) {
            return 0; // Should not happen given constraints n>=2
        }
        return power(2, pathToMaxDepthNode - 1);
    }
};
```